### DISCUSSION TITLE
Rebalancing of Partitions in Kafka Brokers

### DISCUSSION THREAD

**Discussion by Shreya21Mishra** [strimzi/strimzi-kafka-operator#9325] [General]
Hi. I am looking for options to automatically rebalance partitions on Kafka brokers and customise the criteria for rebalancing, other than relying solely on rack awareness
1. Are there any configurable parameters or annotations in Strimzi that enable automatic partition rebalancing among Kafka brokers?
2. Is there any way to set the rebalancing criteria to broker or the disks rather than rackawarness?
3. Also, is there an automatic intra broker rebalancing i.e., rebalancing between the different disks of a broker?

**Comment by ppatierno** [comment]
There is nothing "automated" in the sense that the Strimzi operator looks the cluster for you and decide to rebalance partitions but we have integration with Cruise Control. Did you have a look at the doc? There is an entire chapter on it https://strimzi.io/docs/operators/latest/deploying#cruise-control-concepts-str

**Reply by scholzj** [reply]
Just to add ... if you want, you can easily _script_ this through the `KafkaRebalance` resources if you really want it.

**Reply by Shreya21Mishra** [reply]
@ppatierno and @scholzj  I see there is an annotation in place from Strimzi 
`  annotations:
    strimzi.io/rebalance-auto-approval: "true"`
    Which will automatically approve the proposal and apply it.

1. Is this proposal generated when a new topic is created or when a consumer is instantiated? Basically what is the scenario that triggers this creation?
2. Partition reassignment commands - if I set rebalanceDisk as true, does that mean we wont be able to set any other goals on broker level like ReplicaDistributionGoal etc.

<img width="771" alt="image" src="https://github.com/strimzi/strimzi-kafka-operator/assets/108257512/9ccb704a-aebc-4eb2-ad71-1df73796f3e1">

**Reply by ppatierno** [reply]
Regarding 1 ... did you take a look at documentation? It's explained there https://strimzi.io/docs/operators/latest/deploying#manually_approving_or_rejecting_an_optimization_proposal
Regarding 2 ... that's it. It seems to be a Cruise Control limitation.

**Reply by Shreya21Mishra** [reply]
@ppatierno Yes I did check that out. But it does not mention when the proposal will be created.
https://strimzi.io/docs/operators/latest/deploying#automatically_approving_an_optimization_proposal

usecase:
We want to intra broker rebalancing to happen when the disk utilisation reaches a threshold of 80%. Post that we want to increase the number of brokers.

**Reply by ppatierno** [reply]
I think the documentation on the Cruise Control integration is pretty clear about stating that everything starts with a KafkaRebalance custom resource always created by the user. It triggers the request for a proposal to CC (calling the API) which returns a proposal in the resource status. Then you have to approve it manually to run the rebalance or set the autoapproval to avoid this step.

**Reply by Shreya21Mishra** [reply]
@ppatierno I get that. What I want to know is, say we have a KafkaRebalance CR file where the autoapproval is set to true already present in the cluster. When will the rebalancing be triggered - like upon the addition of a new topic, increasing the partition of an existing topic?

**Reply by ppatierno** [reply]
Just when the proposal is ready. When it's returned by the CC and the KafkaRebalance status is updated with its information.

**Reply by Shreya21Mishra** [reply]
@ppatierno 
If I have a KafkaRebalance already in ready state, will the cruise control automatically detect the any skewness and create the proposal and rebalance the cluster
<img width="946" alt="Screenshot 2023-11-07 at 4 19 45 PM" src="https://github.com/strimzi/strimzi-kafka-operator/assets/108257512/ae1e008b-d774-4671-963c-10c6d614f67b">
For example I already have the above, does this mean it will automatically keep generating proposals and rebalancing the cluster whenever necesary?

**Reply by ppatierno** [reply]
Nope. I think you should read again the doc and pointing here where it's not clear. It also mention a "refresh" annotation in order to make a new proposal from the same KafkaRebalance resource. As I already mentioned, there is no automation on CC generating proposals or doing rebalancing without user interaction.

**Reply by Shreya21Mishra** [reply]
@ppatierno  Can I do partition rebalance between 2 specific brokers?

**Reply by ppatierno** [reply]
This could be a more Cruise Control related question. AFAICS from the Cruise Control wiki here https://github.com/linkedin/cruise-control/wiki/REST-APIs#trigger-a-workload-balance there is just a parameter named `destination_broker_ids` where you can "specify brokers to move replicas to" which is all brokers by default. Not sure it covers what you are asking for.
Said that, it's anyway not configurable via the Strimzi `KafkaRebalance` custom resource, so there is no support for it right now.
